### PR TITLE
refactor(run/markdown-preview): improve local testing and CI flakiness

### LINKED ISSUE #1935: run/markdown-preview/editor: Can successfully make a request to the Renderer failed
Note: #1877 was also for this test, but it was closed more than 10 days ago. So, I didn't mark it flaky.

----

commit: 0cbaa0c436a7925d1185de226fa9659723d56514
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/da43d025-cb2f-48fd-937f-b2d673185702), [Sponge](http://sponge2/da43d025-cb2f-48fd-937f-b2d673185702)
status: failed
<details><summary>Test output</summary><br><pre>Response code 500 (Internal Server Error)
HTTPError: Response code 500 (Internal Server Error)
    at onResponse (node_modules/got/dist/source/as-promise/index.js:142:28)
    at process._tickCallback (internal/process/next_tick.js:68:7)</pre></details>

### PR DESCRIPTION
**Pull request by grayside**
Fixes #1935

### Structure, Test Stability

* Added `Content-Type: application/json` to tests of editor's `/render` endpoint.
* Split `main.js` into `index.js` and `app.js`. This way app.js can be loaded for use with supertest without starting the HTTP server. This can lead to random test failures on port binding conflicts.

### Local Development & Testing

* Reduced number of environment variables needed to run system tests for editor
* Clearer separation in shell code and output for which service is being managed
* Assert `GOOGLE_CLOUD_PROJECT` as a required environment variable

### Documentation

* Improved Cloud Run README documentation on running tests locally.

### Streamlining

* Reduced redundancy in error message output
* Moved in-line test timeout to mocha flag
* Removed test cases which exercise our dependencies rather than functionality in this code
* More concise and consistent labels for test cases with `it()`

### Compatibility & Practices Improvement

* Adjusted Dockerfiles for `CMD ["node", "index.js"]` entrypoint for compatibility with `skaffold debug`

### CODE REVIEW

File: run/README.md [hunk 1]
```diff
@@ -101,9 +101,22 @@ For more Cloud Run samples beyond Node.js, see the main list in the [Cloud Run S
...
+SAMPLE=[SAMPLE_TO_TEST]
+gcloud builds submit --tag gcr.io/$GOOGLE_CLOUD_PROJECT/${SAMPLE}:manual
+CONTAINER_IMAGE=gcr.io/$GOOGLE_CLOUD_PROJECT/${SAMPLE}:manual SERVICE_NAME=${SAMPLE} npm run system-test
+gcloud container images delete gcr.io/$GOOGLE_CLOUD_PROJECT/${SAMPLE}:manual
 ```
+
+## Deploying
+
+```sh
 gcloud builds submit --tag gcr.io/${GOOGLE_CLOUD_PROJECT}/${SAMPLE}
```
Comment (averikitsch) [inline run/README.md:120]: nit: consistency with using ${VAR} vs $VAR

File: run/markdown-preview/editor/index.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+const {app} = require('./app');
+const pkg = require('./package.json');
+const PORT = process.env.PORT || 8080;
+
+app.listen(PORT, () =>
+  console.log(`${pkg.name} listening on port ${PORT}`)
```
Comment (averikitsch) [inline run/markdown-preview/editor/index.js:6]: nit: update name in package.json to markdown-preview-editor

File: .kokoro/build-with-run.sh [hunk 3]
```diff
@@ -44,7 +44,8 @@ gcloud config set project $GOOGLE_CLOUD_PROJECT
 # Version is in the format <PR#>-<GIT COMMIT SHA>.
 # Ensures PR-based triggers of the same branch don't collide if Kokoro attempts
 # to run them concurrently. Defaults to 'latest'.
-export SAMPLE_VERSION="${KOKORO_GIT_COMMIT:-latest}"
+RAW_SAMPLE_VERSION="${KOKORO_GIT_COMMIT:-latest}"
+export SAMPLE_VERSION="${RAW_SAMPLE_VERSION:0:15}"
```
Comment (averikitsch) [inline .kokoro/build-with-run.sh:48]: nit: add comment about shortening the sample version to fit cloud run limitations

File: run/README.md [hunk 4]
```diff
@@ -101,9 +101,22 @@ For more Cloud Run samples beyond Node.js, see the main list in the [Cloud Run S
...
+## Running the Tests
+
+```sh
+# Run unit & integration tests.
+npm test
 
+# Run system tests.
+SAMPLE=[SAMPLE_TO_TEST]
+gcloud builds submit --tag gcr.io/$GOOGLE_CLOUD_PROJECT/${SAMPLE}:manual
+CONTAINER_IMAGE=gcr.io/$GOOGLE_CLOUD_PROJECT/${SAMPLE}:manual SERVICE_NAME=${SAMPLE} npm run system-test
```
Comment (averikitsch) [inline run/README.md:113]:
```suggestion
CONTAINER_IMAGE=gcr.io/$GOOGLE_CLOUD_PROJECT/${SAMPLE}:manual 
gcloud builds submit --tag $CONTAINER_IMAGE
SERVICE_NAME=${SAMPLE} npm run system-test
```

File: run/README.md [hunk 5]
```diff
@@ -101,9 +101,22 @@ For more Cloud Run samples beyond Node.js, see the main list in the [Cloud Run S
...
-## Deploying
+## Running the Tests
+
+```sh
+# Run unit & integration tests.
+npm test
 
+# Run system tests.
+SAMPLE=[SAMPLE_TO_TEST]
+gcloud builds submit --tag gcr.io/$GOOGLE_CLOUD_PROJECT/${SAMPLE}:manual
```
Comment (averikitsch) [inline run/README.md:112]: Do we set GOOGLE_CLOUD_PROJECT anywhere in the README or do we assume they have it set?

File: run/markdown-preview/editor/render.js [hunk 6]
```diff
@@ -43,15 +43,15 @@ const renderRequest = async (markdown) => {
     const clientHeaders = await client.getRequestHeaders();
     serviceRequestOptions.headers['Authorization'] = clientHeaders['Authorization'];
   } catch(err) {
-    throw Error('GoogleAuth server could not respond to request: ', err);
+    throw Error('could not create identity token: ', err);
```
Comment (averikitsch) [inline run/markdown-preview/editor/render.js:46]: nit
```suggestion
    throw Error('could not create an identity token: ', err);
```

File: run/markdown-preview/editor/test/runner.sh [hunk 7]
```diff
@@ -20,38 +20,40 @@ requireEnv() {
...
+echo ":: Deploy the Renderer service"
+echo
 pushd ../renderer
 
 # Version is in the format <PR#>-<GIT COMMIT SHA>.
 # Ensures PR-based triggers of the same branch don't collide if Kokoro attempts
 # to run them concurrently.
-export UPSTREAM_SAMPLE_VERSION="${KOKORO_GIT_COMMIT:-latest}"
-export UPSTREAM_CONTAINER_IMAGE="gcr.io/${GOOGLE_CLOUD_PROJECT}/run-renderer:${UPSTREAM_SAMPLE_VERSION}"
+export UPSTREAM_CONTAINER_IMAGE="gcr.io/${GOOGLE_CLOUD_PROJECT}/run-renderer:${SAMPLE_VERSION:-latest}"
```
Comment (averikitsch) [inline run/markdown-preview/editor/test/runner.sh:35]: Where is SAMPLE_VERSION set or propagating from?

File: run/markdown-preview/editor/test/runner.sh [hunk 8]
```diff
@@ -20,38 +20,40 @@ requireEnv() {
...
 
-pushd ../editor
-
-requireEnv UPSTREAM_CONTAINER_IMAGE
-
 # Assign the Renderer service container image.
-SUFFIX=${KOKORO_BUILD_ID}
-export UPSTREAM_SERVICE_NAME="renderer-${SUFFIX}"
-requireEnv UPSTREAM_SERVICE_NAME
+export UPSTREAM_SERVICE_NAME="renderer-${SUFFIX:-manual}"
```
Comment (averikitsch) [inline run/markdown-preview/editor/test/runner.sh:43]: Where is SUFFIX set?

File: run/markdown-preview/editor/test/runner.sh [hunk 9]
```diff
@@ -20,38 +20,40 @@ requireEnv() {
...
-export UPSTREAM_SAMPLE_VERSION="${KOKORO_GIT_COMMIT:-latest}"
-export UPSTREAM_CONTAINER_IMAGE="gcr.io/${GOOGLE_CLOUD_PROJECT}/run-renderer:${UPSTREAM_SAMPLE_VERSION}"
+export UPSTREAM_CONTAINER_IMAGE="gcr.io/${GOOGLE_CLOUD_PROJECT}/run-renderer:${SAMPLE_VERSION:-latest}"
 
 # Build the Renderer service
 set -x
 gcloud builds submit --tag="${UPSTREAM_CONTAINER_IMAGE}"
 set +x
 
-pushd ../editor
```
Comment (averikitsch) [inline run/markdown-preview/editor/test/runner.sh:40]: Do we not need to be in the editor folder to access test/deploy.sh?

File: .kokoro/build-with-run.sh (hunk 3 above)
Comment (grayside) [reply .kokoro/build-with-run.sh:48]: This change is not related to Cloud Run, but was a blind experiment to see if GCR had a limitation. It seemed to help, 🤷

File: run/README.md (hunk 5 above)
Comment (grayside) [reply run/README.md]: Probably assumed. But as the README is not a tutorial, I felt it was reasonable to rely on gcloud to throw an error. I can add something like the SAMPLE var just above if you think that would help.

File: run/markdown-preview/editor/test/runner.sh (hunk 7 above)
Comment (grayside) [reply run/markdown-preview/editor/test/runner.sh:35]: .kokoro/build-with-run.sh

File: run/markdown-preview/editor/test/runner.sh (hunk 8 above)
Comment (grayside) [reply run/markdown-preview/editor/test/runner.sh:43]: .kokoro/build-with-run.sh

File: run/markdown-preview/editor/test/runner.sh (hunk 9 above)
Comment (grayside) [reply run/markdown-preview/editor/test/runner.sh:40]: This is calling `run/markdown-preview/renderer/test/deploy.sh`.

File: run/README.md [hunk 10]
```diff
@@ -101,9 +101,23 @@ For more Cloud Run samples beyond Node.js, see the main list in the [Cloud Run S
...
+gcloud builds submit --tag $CONTAINER_IMAGE
+SERVICE_NAME=${SAMPLE} npm run system-test
+gcloud container images delete gcr.io/$GOOGLE_CLOUD_PROJECT/${SAMPLE}:manual
 ```
+
+## Deploying
+
+```sh
 gcloud builds submit --tag gcr.io/${GOOGLE_CLOUD_PROJECT}/${SAMPLE}
 gcloud run deploy $SAMPLE \
```
Comment (grayside) [inline run/README.md:122]:
```suggestion
gcloud run deploy ${SAMPLE} \
```

**Pull request merged**